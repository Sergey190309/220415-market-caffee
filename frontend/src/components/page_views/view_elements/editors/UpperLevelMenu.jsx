import React, { useState, useEffect, useContext, Fragment } from 'react'
import PropTypes from 'prop-types'
import {
  Popup,
  Menu
  // Dropdown
} from 'semantic-ui-react'
import { useTranslation } from 'react-i18next'

import {
  UPPER_ELEMENT_ADD_ABOVE, UPPER_ELEMENT_ADD_BELOW,
  UPPER_ELEMENT_DELETE
} from '../../../../redux/constants/menuKeys'
import { LandingContext, ElementSwitcherContext } from '../../../../context'
import { neutralColor, warningColor } from '../../../../utils/colors'
import { ContextMenuItem } from '../../../items/menu_items/ContextMenuItem'

// import { deviceSelector } from '../../../../redux/slices'

/**
 * To handle upper level elements I need:
 *    - view id (landing or something)
 *    - serial index - id, string in my case.
 *    - upper level type - choose from upperLevelTypeMenu (Header,
 *        Footer, VBlock, HBlock)
 *    - upper level subtype - from upperLevelSubtypeMenu (txt, pix).
 * Functions to handle upper level elements would be here. They are:
 *    - addUpperLevelElement - provide context to type and subtype menus
 *    - deleteUpperLevelElement - direct dispatch here in onClickHandler.
 */

export const addUppreLevelElement = (viewId, upperElementIndex, type = 'header', subtype = 'txt') => {
  console.log('UpperLevelMenu:\n addUppreLevelElement',
    '\n  viewId ->', viewId,
    '\n  upperElementIndex ->', upperElementIndex,
    '\n  type ->', type,
    '\n  subtype ->', subtype
  )
}

const UpperLevelMenu = ({
  // isOpened,
  context,
  setMenuOpened,
  setUpperLevelTypeMenuOpened,
  addBelow
}) => {
  const { t } = useTranslation('context')
  const [opened, setOpened] = useState(false)
  const [menuStructure, setMenuStructure] = useState([])

  const { componentName: viewId } = useContext(LandingContext)
  const { recordsId } = useContext(ElementSwitcherContext)

  // const { editable } = useSelector(deviceSelector)
  useEffect(() => {
    // console.log('UpperLevelMenu:\n useEffect[isMenuOpened]',
    //   '\n  isMenuOpened ->', isMenuOpened
    // )
    setOpened(true)
    setMenuStructure([
      {
        name: UPPER_ELEMENT_ADD_ABOVE,
        icon: { name: 'angle up', color: neutralColor },
        content: t('1LE.addAbove')
      },
      {
        name: UPPER_ELEMENT_ADD_BELOW,
        icon: { name: 'angle down', color: neutralColor },
        content: t('1LE.addBelow')
      },
      {
        name: UPPER_ELEMENT_DELETE,
        icon: { name: 'delete', color: warningColor },
        content: t('1LE.DeleteElement')
      }
    ])
    return () => {
      // console.log('UpperLevelMenu:\n useEffect[], clean up',
      //   '\n  isMenuOpened ->', isOpened
      // )
      setMenuOpened(false)
    }
  }, [])

  const onClickHandler = (event, { name }) => {
    event.preventDefault()
    switch (name) {
      case UPPER_ELEMENT_ADD_ABOVE:
        console.log('UpperLevelMenu:\n onClickHandler',
          '\n  name ->', name)
        addBelow(false)
        setUpperLevelTypeMenuOpened(true)
        // addUppreLevelElement(viewId, +recordsId.split('_')[0])
        break
      case UPPER_ELEMENT_ADD_BELOW:
        console.log('UpperLevelMenu:\n onClickHandler',
          '\n  name ->', name)
        // addUppreLevelElement(viewId, +recordsId.split('_')[0] + 1)
        addBelow(true)
        setUpperLevelTypeMenuOpened(true)
        break
      case UPPER_ELEMENT_DELETE:
        console.log('UpperLevelMenu:\n onClickHandler',
          '\n  name ->', name,
          '\n  viewId ->', viewId,
          '\n  recordsId ->', recordsId,
          '\n  index ->', recordsId.split('_')[0]
        )
        break

      default:
        break
    }
    setMenuOpened(false)
  }

  return (
    <Fragment>
      <Popup
        data-testid='Popup'
        hoverable
        wide='very'
        context={context}
        open={opened}
        // open={opened && editable}
        onClose={() => {
          // console.log('UpperLevelMenu:\n onClose')
          // setOpened(false)
          setMenuOpened(false)
        }}
      >
        <Menu
          vertical
          compact
        >
          {menuStructure.map((item, index) => (
            <ContextMenuItem
              key={index} {...item} onClick={onClickHandler}
            />
          ))}
        </Menu>
      </Popup>
    </Fragment>
  )
}

UpperLevelMenu.defaultProps = {
  isOpened: true,
  context: {},
  setMenuOpened: () => { },
  setUpperLevelTypeMenuOpened: () => { },
  addBelow: () => { }
}

UpperLevelMenu.propTypes = {
  isOpened: PropTypes.bool.isRequired,
  context: PropTypes.object.isRequired,
  setMenuOpened: PropTypes.func.isRequired,
  setUpperLevelTypeMenuOpened: PropTypes.func.isRequired,
  addBelow: PropTypes.func.isRequired
}

export default UpperLevelMenu
