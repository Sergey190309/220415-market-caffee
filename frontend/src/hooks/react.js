import {useState, useEffect, useRef, useCallback } from 'react'

export const useAppState = useState
export const useAppEffect = useEffect
export const useAppRef = useRef
export const useAppCallback = useCallback
