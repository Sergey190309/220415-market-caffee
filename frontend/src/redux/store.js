import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
// import reducer from './reducers';
import reducer from './slices'
import rootSaga from './saga'
// import { rootReducer } from './reducers';

const initialState = {}
const sagaMiddleware = createSagaMiddleware()

const middleware = [...getDefaultMiddleware({ thunk: false }), sagaMiddleware]
// const middleware = [sagaMiddleware, thunk];

const store = configureStore({
  reducer,
  middleware,
  initialState
})

sagaMiddleware.run(rootSaga)

export default store
