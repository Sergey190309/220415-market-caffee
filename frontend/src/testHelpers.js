import React from 'react'

import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

// import store from './redux/store'

export const testRender = (jsx, { store, ...otherOpts }) => {
  return render(
    <Provider store={store}>
      {jsx}
    </Provider>,
    otherOpts
  )
}
// export function testRender (jsx, { store, ...otherOpts }) {
//   return render(<Provider store={store}>{jsx}</Provider>, otherOpts)
// }
