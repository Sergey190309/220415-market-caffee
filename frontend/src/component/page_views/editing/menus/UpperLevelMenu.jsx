import React from 'react'
import { useAppEffect, useAppState } from '../../../../hooks/react'
import { useAppSelector } from '../../../../hooks/reactRedux'
import PropTypes from 'prop-types'
import { Menu } from '@mui/material'
import { CancelOutlined, DeleteOutline, KeyboardArrowDown, KeyboardArrowUp } from '@mui/icons-material'
import { useTranslation } from 'react-i18next'

import { deviceSelector } from '../../../../redux/slices'
import ContextMenuItem from './ContextMenuItem'
import * as CL from '../../../../constants/colors'
import UpperLevelType from './UpperLevelType'

const UpperLevelMenu = ({ upperLevelMenu, upperLevelMenuCloseHandler }) => {
  /**
   * add above
   * add below
   * delete
   * cancel
   */
  const [opened, setOpened] = useAppState(false)
  const [upperLevelType, setUpperLevelType] = useAppState(null)

  const { editable } = useAppSelector(deviceSelector)

  const { t } = useTranslation('menus')

  useAppEffect(() => {
    setOpened(upperLevelMenu !== null)
  }, [upperLevelMenu])


  const onAddElementHandler = (event, above) => {
    // console.log('UpperLevelMenu>onAddElementHandler, above ->', above)
    event.preventDefault()
    if (editable) {
      // console.log('ContextMenu>onUpperElementHandler, event ->', event)
      setUpperLevelType({
        mouseX: event.clientX + 2,
        mouseY: event.clientY - 6,
      })
    }
    // setOpened(false)
  }
  const onDeleteHandler = () => {
    // console.log('UpperLevelMenu>onDeleteHandler')
    upperLevelMenuCloseHandler(true)
    setOpened(false)
  }
  const onCancelHandler = () => {
    // console.log('UpperLevelMenu>onCancelHandler')
    setOpened(false)
  }
  const onCloseHandler = () => {
    upperLevelMenuCloseHandler()
    setOpened(false)
  }

  const upperLevelTypeCloseHandler = closeAll => {
    setUpperLevelType(null)
    if (closeAll) {
      upperLevelMenuCloseHandler(closeAll)
    }
  }
  // console.log('UpperLevelMenu, render, upperLevelType ->', upperLevelType)

  return (
    <>
      <Menu
        open={opened}
        anchorReference={'anchorPosition'}
        anchorPosition={upperLevelMenu !== null ? {
          top: upperLevelMenu.mouseY, left: upperLevelMenu.mouseX
        } : undefined}
        onClose={onCloseHandler}
        PaperProps={{
          sx: {
            bgcolor: CL.bodyBackground
          }
        }}
      >
        <ContextMenuItem
          title={t('upperLevelMenu.addAbove')}
          Icon={KeyboardArrowUp}
          // onClick={onAddElementHandler} />
          onClick={(e) => { onAddElementHandler(e, true) }} />
        <ContextMenuItem
          title={t('upperLevelMenu.addBelow')}
          Icon={KeyboardArrowDown}
          onClick={(e) => { onAddElementHandler(e, false) }} />
        <ContextMenuItem
          title={t('upperLevelMenu.delete')}
          Icon={DeleteOutline}
          onClick={onDeleteHandler} />
        <ContextMenuItem
          title={t('upperLevelMenu.cancel')}
          Icon={CancelOutlined}
          onClick={onCancelHandler} />
      </Menu>
      {upperLevelType === null ? null :
        <UpperLevelType
          upperLevelType={upperLevelType}
          upperLevelTypeCloseHandler={upperLevelTypeCloseHandler}
        />
      }
    </>
  )
}

UpperLevelMenu.defaultProps = {
  upperLevelMenu: {},
  upperLevelMenuCloseHandler: () => { }
}
UpperLevelMenu.propTypes = {
  upperLevelMenu: PropTypes.object,
  upperLevelMenuCloseHandler: PropTypes.func.isRequired
}

export default UpperLevelMenu