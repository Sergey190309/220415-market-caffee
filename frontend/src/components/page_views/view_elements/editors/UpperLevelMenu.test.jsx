import React from 'react'
import { screen, waitFor } from '@testing-library/react'

import { ElementSwitcherProvider, LandingProvider }
  from '../../../../context'
import store from '../../../../redux/store'
import { makeTestStore, testRender, elementFunc } from '../../../../testHelpers'
import UpperLevelMenu, { upperLevelMenu } from './UpperLevelMenu'
import userEvent from '@testing-library/user-event'

jest.mock('../../../../testHelpers', () => ({
  ...jest.requireActual('../../../../testHelpers'),
  elementFunc: jest.fn()
}))

describe('UpperLevelMenu', () => {
  const testProps = {
    context: {},
    setMenuOpened: jest.fn(),
    setUpperLevelTypeMenuOpened: jest.fn(),
    addBelow: jest.fn()
  }
  const mockContext = {
    recordsId: { recordsId: '01_vblock_txt' },
    componentName: { componentName: 'landing' }
  }
  const renderReduxContext = (
    actualProps,
    testStore = makeTestStore(store),
    context = mockContext
  ) => {
    // console.log('ParagraphContextMenu.test:\n renderReduxContext',
    //   '\n  index ->', index)
    testRender(
      <LandingProvider value={context.componentName} >
        <ElementSwitcherProvider
          value={context.recordsId} >
          <UpperLevelMenu {...actualProps} />
        </ElementSwitcherProvider>
      </LandingProvider>,
      { store: testStore }
    )
  }
  describe('appearance', () => {
    test('render (snapshot)', async () => {
      const activeProps = { ...testProps }
      await waitFor(() => { renderReduxContext(activeProps, store) })
      const popup = screen.getByTestId('Popup')
      expect(popup).toMatchSnapshot()
      expect(popup.children[0].children)
        .toHaveLength(upperLevelMenu(jest.fn()).length)
    })
  })
  describe('actions', () => {
    test('1LE.addAbove ', async () => {
      const activeProps = { ...testProps }
      await waitFor(() => { renderReduxContext(activeProps, store) })
      const addAbove = screen.getByText('1LE.addAbove')
      userEvent.click(addAbove)
      expect(elementFunc).toHaveBeenCalledTimes(1)
      expect(elementFunc)
        .toHaveBeenCalledWith({ name: 'UPPER_ELEMENT_ADD_ABOVE' })
      expect(activeProps.addBelow).toHaveBeenCalledTimes(1)
      expect(activeProps.addBelow).toHaveBeenCalledWith(false)
      expect(activeProps.setUpperLevelTypeMenuOpened)
        .toHaveBeenCalledTimes(1)
      expect(activeProps.setUpperLevelTypeMenuOpened)
        .toHaveBeenCalledWith(true)
      expect(activeProps.setMenuOpened).toHaveBeenCalledTimes(1)
      expect(activeProps.setMenuOpened).toHaveBeenCalledWith(false)
    })
    test('1LE.addBelow ', async () => {
      const activeProps = { ...testProps }
      await waitFor(() => { renderReduxContext(activeProps, store) })
      const addBelow = screen.getByText('1LE.addBelow')
      userEvent.click(addBelow)
      expect(elementFunc).toHaveBeenCalledTimes(1)
      expect(elementFunc)
        .toHaveBeenCalledWith({ name: 'UPPER_ELEMENT_ADD_BELOW' })
      expect(activeProps.addBelow).toHaveBeenCalledTimes(1)
      expect(activeProps.addBelow).toHaveBeenCalledWith(true)
      expect(activeProps.setUpperLevelTypeMenuOpened)
        .toHaveBeenCalledTimes(1)
      expect(activeProps.setUpperLevelTypeMenuOpened)
        .toHaveBeenCalledWith(true)
      expect(activeProps.setMenuOpened).toHaveBeenCalledTimes(1)
      expect(activeProps.setMenuOpened).toHaveBeenCalledWith(false)
    })
    test('1LE.DeleteElement ', async () => {
      const activeProps = { ...testProps }
      await waitFor(() => { renderReduxContext(activeProps, store) })
      const DeleteElement = screen.getByText('1LE.DeleteElement')
      userEvent.click(DeleteElement)
      expect(elementFunc).toHaveBeenCalledTimes(1)
      expect(elementFunc)
        .toHaveBeenCalledWith({
          index: mockContext.recordsId.recordsId.split('_')[0],
          name: 'UPPER_ELEMENT_DELETE',
          recordsId: mockContext.recordsId.recordsId,
          viewId: mockContext.componentName.componentName
        })
      expect(activeProps.setMenuOpened).toHaveBeenCalledTimes(1)
      expect(activeProps.setMenuOpened).toHaveBeenCalledWith(false)
      // const popup = screen.getByTestId('Popup')
      // screen.debug(popup)
    })
  })
})
