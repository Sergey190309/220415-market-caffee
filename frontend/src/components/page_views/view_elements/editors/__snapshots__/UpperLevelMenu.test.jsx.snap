// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpperLevelMenu appearance render (snapshot) 1`] = `
<div
  class="ui top left very wide popup transition visible"
  data-testid="Popup"
  style="position: initial;"
>
  <div
    class="ui compact vertical menu"
  >
    <a
      class="item sc-bdfBQB eVLmMB"
    >
      <i
        aria-hidden="true"
        class="olive angle up icon"
      />
      1LE.addAbove
    </a>
    <a
      class="item sc-bdfBQB eVLmMB"
    >
      <i
        aria-hidden="true"
        class="olive angle down icon"
      />
      1LE.addBelow
    </a>
    <a
      class="item sc-bdfBQB eVLmMB"
    >
      <i
        aria-hidden="true"
        class="orange delete icon"
      />
      1LE.DeleteElement
    </a>
  </div>
</div>
`;
